// unleash-client/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppliedJitter = exports.generateHashOfConfig = exports.generateInstanceId = exports.safeName = exports.resolveContextValue = exports.createFallbackFunction = void 0;
const os_1 = require("os");
const murmurHash3 = require("murmurhash3js");
function createFallbackFunction(name, context, fallback) {
    if (typeof fallback === 'function') {
        return () => fallback(name, context);
    }
    if (typeof fallback === 'boolean') {
        return () => fallback;
    }
    return () => false;
}
exports.createFallbackFunction = createFallbackFunction;
function resolveContextValue(context, field) {
    var _a, _b, _c, _d;
    return (_b = (_a = context[field]) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : (_d = (_c = context.properties) === null || _c === void 0 ? void 0 : _c[field]) === null || _d === void 0 ? void 0 : _d.toString();
}
exports.resolveContextValue = resolveContextValue;
function safeName(str = '') {
    return str.replace(/\//g, '_');
}
exports.safeName = safeName;
function generateInstanceId(instanceId) {
    if (instanceId) {
        return instanceId;
    }
    let info;
    try {
        info = (0, os_1.userInfo)();
    }
    catch (e) {
        // unable to read info;
    }
    const prefix = info
        ? info.username
        : `generated-${Math.round(Math.random() * 1000000)}-${process.pid}`;
    return `${prefix}-${(0, os_1.hostname)()}`;
}
exports.generateInstanceId = generateInstanceId;
function generateHashOfConfig(o) {
    const oAsString = JSON.stringify(o);
    return murmurHash3.x86.hash128(oAsString);
}
exports.generateHashOfConfig = generateHashOfConfig;
function getAppliedJitter(jitter) {
    const appliedJitter = Math.random() * jitter;
    return Math.random() < 0.5 ? -appliedJitter : appliedJitter;
}
exports.getAppliedJitter = getAppliedJitter;
